### ISSUE TITLE
20140508: Screen Activates Before App.

### ISSUE THREAD

**Issue by ddantzler** [lzanita09/NotificationPeekPort#11]
In the nightly build, the screen turns on before the app when a notification arrives on my Nexus 5, Root, Xposed. v1 does not appear to have this problem.

**Comment by lzanita09** [comment]
Do you mean the screen turns on itself even there is no notification coming?

**Comment by ddantzler** [comment]
The notification is coming. The screen lights up and I see the lock screen briefly. I see the notification in the header. Then, Peek appears.

**Comment by ddantzler** [comment]
Not a problem in today's build.

**Comment by ddantzler** [comment]
Oops, spoke too soon. If gyroscope sensor is turned off, the issue reappears. So, the screen will awake. Lock screen appears. Notification is popping into the status bar. Then, a split second later, the Peek app appears.

**Comment by lzanita09** [comment]
Does it only happen with gyro off?

**Comment by ddantzler** [comment]
Yes, based on my testing. 

There is a separate issue that occurs also where at time out, when the screen is supposed to turn off, the Peek activity disappears a split second before the screen turns off and locks. This means that whatever app is behind Peek is revealed for a split instant.

This occurs regardless of sensor settings based on my testing.

**Comment by ddantzler** [comment]
More details: Seems to affect every Hangouts notification but not Gmail.

**Comment by lzanita09** [comment]
What about the other apps?

**Comment by ddantzler** [comment]
Will continue testing and follow up.

**Comment by ddantzler** [comment]
Having difficulty identifying other apps. Seems intermittently effecting Gmail actually, as well as Email. Seems to effect Hangouts often. Primarily an issue when Gyro is disabled. Gyro seems to mitigate it, though can happen with Gyro too, since the phone is in the hand. To me it seems there is a time delay that is set a bit too short on both lighting of the screen, and maybe too long on killing the screen, since Peek seems to close before the screen is turned off. I'll try to capture a video tonight if possible if that will help you.

**Comment by ddantzler** [comment]
"Since the phone is in the hand," I mean if you're trying to test by sending yourself a notification. IE: SMS from Voice # to main #. So if Gyro is on, its still a problem if the screen is off and actively moving in the hand. Screen lights up for a split second and then Peek appears.

**Comment by lzanita09** [comment]
Yeah a video will definitely help. Thanks.

**Comment by ddantzler** [comment]
So, I'm noticing that my testing method for this might have been the problem. I'll keep testing and try to get more details for you. 

At this time, I know that you can replicate it by: 1) Start Google Voice. 2) Send an SMS to yourself. 3) Quickly lock the screen (before the SMS arrives). 4) View the behavior when the SMS arrives.

If you're like me, the screen will light up on the lock screen and then NotificationPeek will start.

I have attached a video of the other error, which is more pervasive. Watch how the timeout happens. You'll see the application in the background before the screen locks.

https://www.youtube.com/watch?v=CsD_aqIZShI

The setting I have enabled in the app are as follows:
Show Clock
Notification Peek Timeout: 10 sec
Use Proximity/Light Sensor

Other settings are off.

**Comment by ddantzler** [comment]
Here is a better video. In this video, I generate an SMS using Voice. I immediately lock the phone on sending with the power button. The screen lights and NotificationPeek appears. I wait for the timeout. A flash occurs before the screen turns off. Finally, I use the power button to turn on the device and stop the screen recording.

http://youtu.be/8S5D_ZKhceI

**Comment by lzanita09** [comment]
Currently the code is written like this, upon receiving notification, it first launch an Activity with the Peek View, and then after 300 ms, turn on the screen. And after 10 sec in this case, it turns off the screen, and then 3 seconds later, remove the view. 

The weird thing is that even thought I have the same device I couldn't reproduce the issue. I assume that you're using stock rom, do you use ART or Dalvik?

**Comment by ddantzler** [comment]
Stock ROM, rooted, Xposed framework with GravityBox. I reviewed the code and saw the delays. The only other thing that may be configured weird for me in GravityBox is a custom screen off animation (fade) instead of the stock CRT type. I also made sure I have no other applications with notification access, or accessibility access out of the ordinary. I will keep testing.

**Comment by lzanita09** [comment]
I'm not familiar with Xposed or GravityBox, so I can't tell whether they affect the delay. Thank you for your testing and report.

**Comment by ddantzler** [comment]
Okay, turning off GravityBox or changing the screen off animation to stock in GravityBox fixes the screen off problem. The screen on problem seems to only appear when turning the screen off and receiving a notification within a very short time frame. That's a very rare case, except with the unique test that I am using.

**Comment by ddantzler** [comment]
Also, I notice that NotificationPeek appears in recent tasks after notification received.

**Comment by ddantzler** [comment]
Donated you via your email address on PayPal. Hope you get it. Thanks for your work on this app.

**Comment by lzanita09** [comment]
That's correct, because currently I use an Activity to display the View, it seems to be easier to control than using system alert dialog. And the trade off is that you can see the Activity in recent tasks.

Yes I just got it, thank you!

**Comment by ddantzler** [comment]
Not only is it in Recent, sometimes if a notification appears and you tap in the black area to unlock the device, the NotificationPeek settings window is what you see.

**Comment by lzanita09** [comment]
Do you mean the settings page is launched automatically behind the Peek View?

**Comment by ddantzler** [comment]
Yes. Notification received. Peek View appears. Tap the screen (not the notification). Settings page appears. Would think that the home screen should appear.

**Comment by lzanita09** [comment]
The home screen should be the correct behavior, I'll take a look at the code.

**Comment by ddantzler** [comment]
Meanwhile will try to get a video, but I am having a hard time isolating it.

**Comment by lzanita09** [comment]
It's fine, I'll just see if I can reproduce it, you have described the issue quite clearly.
